File: 3598.longest-common-prefix-between-adjacent-strings-after-removals.go
#
# @lc app=leetcode id=3598 lang=golang
#
# [3598] Longest Common Prefix Between Adjacent Strings After Removals
#

# @lc code=start
func longestCommonPrefix(words []string) []int {
	maxx := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}

	lcp := func(a, b string) int {
		minl := len(a)
		if minl > len(b) {
			minl = len(b)
		}
		for i := 0; i < minl; i++ {
			if a[i] != b[i] {
				return i
			}
		}
		return minl
	}

	n := len(words)
	if n <= 1 {
		return []int{0}
	}

	m := n - 1
	adj := make([]int, m)
	for i := 0; i < m; i++ {
		adj[i] = lcp(words[i], words[i+1])
	}

	const LOG = 18
	st := make([][]int, m)
	for i := range st {
		st[i] = make([]int, LOG)
		st[i][0] = adj[i]
	}

	for j := 1; j < LOG; j++ {
		for ii := 0; ii+(1<<j) <= m; ii++ {
			st[ii][j] = maxx(st[ii][j-1], st[ii+(1<<(j-1))][j-1])
		}
	}

	logg := make([]int, m+2)
	logg[0] = 0
	logg[1] = 0
	for i := 2; i <= m; i++ {
		logg[i] = logg[i>>1] + 1
	}

	qmax := func(L, R int) int {
		if L > R {
			return 0
		}
		ln := R - L + 1
		k := logg[ln]
		return maxx(st[L][k], st[R-(1<<k)+1][k])
	}

	ans := make([]int, n)
	for i := 0; i < n; i++ {
		orig := 0
		if i == 0 {
			orig = qmax(1, m-1)
		} else if i == n-1 {
			orig = qmax(0, m-2)
		} else {
			orig = maxx(qmax(0, i-2), qmax(i+1, m-1))
		}
		newp := 0
		if i > 0 && i < n-1 {
			newp = lcp(words[i-1], words[i+1])
		}
		ans[i] = maxx(orig, newp)
	}

	return ans
}
# @lc code=end

File: 3598.longest-common-prefix-between-adjacent-strings-after-removals.java
#
# @lc app=leetcode id=3598 lang=java
#
# [3598] Longest Common Prefix Between Adjacent Strings After Removals
#

# @lc code=start
class Solution {
    public int[] longestCommonPrefix(String[] words) {
        int n = words.length;
        int[] ans = new int[n];
        if (n <= 1) {
            return ans;
        }
        int m = n - 1;
        int[] prefix = new int[m];
        for (int j = 0; j < m; j++) {
            prefix[j] = getLCP(words[j], words[j + 1]);
        }
        int[] premax = new int[m];
        premax[0] = prefix[0];
        for (int j = 1; j < m; j++) {
            premax[j] = Math.max(premax[j - 1], prefix[j]);
        }
        int[] sufmax = new int[m];
        sufmax[m - 1] = prefix[m - 1];
        for (int j = m - 2; j >= 0; j--) {
            sufmax[j] = Math.max(sufmax[j + 1], prefix[j]);
        }
        for (int i = 0; i < n; i++) {
            int max_kept = 0;
            if (i >= 2) {
                max_kept = Math.max(max_kept, premax[i - 2]);
            }
            if (i <= n - 3) {
                max_kept = Math.max(max_kept, sufmax[i + 1]);
            }
            int new_lcp = 0;
            if (i >= 1 && i <= n - 2) {
                new_lcp = getLCP(words[i - 1], words[i + 1]);
            }
            ans[i] = Math.max(max_kept, new_lcp);
        }
        return ans;
    }
    
    private int getLCP(String a, String b) {
        int len = Math.min(a.length(), b.length());
        for (int k = 0; k < len; k++) {
            if (a.charAt(k) != b.charAt(k)) {
                return k;
            }
        }
        return len;
    }
}
# @lc code=end